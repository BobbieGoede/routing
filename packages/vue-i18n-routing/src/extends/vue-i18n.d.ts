import type { LocaleObject } from '../types'
import type { ComputedRef } from 'vue'

declare module 'vue-i18n' {
  export interface ComposerCustom {
    locales: ComputedRef<string[] | LocaleObject[]>
  }
}
