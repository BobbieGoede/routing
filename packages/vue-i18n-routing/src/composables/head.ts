import { isBoolean, isArray } from '@intlify/shared'
import { ref, watchEffect, isVue3, onUnmounted } from 'vue-demi'
import { useRoute, useRouter } from '@intlify/vue-router-bridge'
import { useI18n } from '@intlify/vue-i18n-bridge'
import { getRouteBaseName, switchLocalePath, localeRoute } from './routing'
import { getLocale, getNormalizedLocales, warn, inBrowser, toRawRoute } from '../utils'
import { DEFAULT_LOCALE, DEFAULT_STRATEGY, STRATEGIES } from '../constants'

import type { Ref } from 'vue-demi'
import type {
  Router,
  VueRouter,
  RouteLocationNormalizedLoaded,
  RouteLocationNormalized,
  Route
} from '@intlify/vue-router-bridge'
import type { I18nHeadOptions, I18nHeadMetaInfo, I18nRoutingOptions, ComposableOptions, MetaAttrs } from './types'
import type { LocaleObject } from '../types'

/**
 * Generate SEO head meta information
 *
 * @param options - An options, see about details {@link I18nHeadOptions}
 *
 * @returns Genereated SEO head meta information
 */
export function useI18nHead({
  addDirAttribute = false,
  addSeoAttributes = false,
  strategy = DEFAULT_STRATEGY,
  defaultLocale = DEFAULT_LOCALE,
  route = useRoute(),
  router = useRouter(),
  i18n = useI18n()
}: Pick<I18nRoutingOptions, 'strategy' | 'defaultLocale'> &
  ComposableOptions &
  I18nHeadOptions = {}): Ref<I18nHeadMetaInfo> {
  type R = Router | VueRouter
  const _router = router as R

  const _defaultLocale = defaultLocale || _router.__defaultLocale
  const _strategy = strategy || _router.__strategy

  const metaObject: Ref<I18nHeadMetaInfo> = ref({
    htmlAttrs: {},
    link: [],
    meta: []
  })

  function cleanMeta() {
    metaObject.value = {
      htmlAttrs: {},
      link: [],
      meta: []
    }
  }

  function updateMeta(_route: RouteLocationNormalizedLoaded | Route) {
    if (i18n.locales == null || i18n.__baseUrl == null) {
      return
    }
    const locale = getLocale(i18n)
    const currentLocale = getNormalizedLocales(i18n.locales.value).find(l => l.code === locale) || {
      code: locale
    }
    const currentLocaleIso = currentLocale.iso
    const currentLocaleDir = currentLocale.dir || (router as R).__defaultDirection

    // Adding Direction Attribute
    if (addDirAttribute) {
      metaObject.value.htmlAttrs!.dir = currentLocaleDir
    }

    // Adding SEO Meta
    if (addSeoAttributes && locale && i18n.locales) {
      if (currentLocaleIso) {
        metaObject.value.htmlAttrs!.lang = currentLocaleIso
      }

      const locales = i18n.locales.value as LocaleObject[]
      addHreflangLinks(locales, i18n.__baseUrl, metaObject.value.link!, {
        defaultLocale: _defaultLocale,
        strategy: _strategy,
        route,
        router,
        i18n
      })
      addCanonicalLinks(i18n.__baseUrl, metaObject.value.link!, addSeoAttributes, { route: _route, router, i18n })
      addCurrentOgLocale(currentLocale, currentLocaleIso, metaObject.value.meta!)
      addAlternateOgLocales(locales, currentLocaleIso, metaObject.value.meta!)
    }
  }

  if (inBrowser) {
    if (isVue3) {
      const stop = watchEffect(() => {
        cleanMeta()
        updateMeta(toRawRoute(_router.currentRoute))
      })
      onUnmounted(() => stop())
    } else {
      /**
       * NOTE:
       * In vue 2 + `@vue/compoistion-api`, useRoute (`$route`) cannot be watched.
       * For this reason, use `afterEach` to work around it.
       */
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const handler = _router.afterEach(
        (to: Route | RouteLocationNormalized, from: Route | RouteLocationNormalized) => {
          cleanMeta()
          updateMeta(to)
        }
      )
      onUnmounted(() => handler())
      updateMeta(route as Route)
    }
  } else {
    updateMeta(toRawRoute(_router.currentRoute))
  }

  return metaObject
}

function addHreflangLinks(
  locales: LocaleObject[],
  baseUrl: string,
  link: MetaAttrs,
  options: Pick<I18nRoutingOptions, 'strategy' | 'defaultLocale'> & ComposableOptions
) {
  if (options.strategy === STRATEGIES.NO_PREFIX) {
    return
  }

  const localeMap = new Map()
  for (const locale of locales) {
    const localeIso = locale.iso

    if (!localeIso) {
      warn('Locale ISO code is required to generate alternate link')
      continue
    }

    const [language, region] = localeIso.split('-')

    if (language && region && (locale.isCatchallLocale || !localeMap.has(language))) {
      localeMap.set(language, locale)
    }

    localeMap.set(localeIso, locale)
  }

  for (const [iso, mapLocale] of localeMap.entries()) {
    const localePath = switchLocalePath(mapLocale.code, options)
    if (localePath) {
      link.push({
        hid: `i18n-alt-${iso}`,
        rel: 'alternate',
        href: toAbsoluteUrl(localePath, baseUrl),
        hreflang: iso
      })
    }
  }

  if (options.defaultLocale) {
    const localePath = switchLocalePath(options.defaultLocale, options)
    if (localePath) {
      link.push({
        hid: 'i18n-xd',
        rel: 'alternate',
        href: toAbsoluteUrl(localePath, baseUrl),
        hreflang: 'x-default'
      })
    }
  }
}

function addCanonicalLinks(
  baseUrl: string,
  link: MetaAttrs,
  seoAttributesOptions: I18nHeadOptions['addSeoAttributes'],
  options: ComposableOptions
) {
  const { route } = options
  const currentRoute = localeRoute(
    {
      ...(route as any), // eslint-disable-line @typescript-eslint/no-explicit-any
      name: getRouteBaseName(route as any, options) // eslint-disable-line @typescript-eslint/no-explicit-any
    },
    undefined,
    options
  )

  if (currentRoute) {
    let href = toAbsoluteUrl(currentRoute.path, baseUrl)

    const canonicalQueries = (!isBoolean(seoAttributesOptions) && seoAttributesOptions!.canonicalQueries) || []

    if (canonicalQueries.length) {
      const currentRouteQueryParams = currentRoute.query
      const params = new URLSearchParams()
      for (const queryParamName of canonicalQueries) {
        if (queryParamName in currentRouteQueryParams) {
          const queryParamValue = currentRouteQueryParams[queryParamName]

          if (isArray(queryParamValue)) {
            queryParamValue.forEach(v => params.append(queryParamName, v || ''))
          } else {
            params.append(queryParamName, queryParamValue || '')
          }
        }
      }

      const queryString = params.toString()

      if (queryString) {
        href = `${href}?${queryString}`
      }
    }

    link.push({
      hid: 'i18n-can',
      rel: 'canonical',
      href
    })
  }
}

function addCurrentOgLocale(currentLocale: LocaleObject, currentLocaleIso: string | undefined, meta: MetaAttrs) {
  const hasCurrentLocaleAndIso = currentLocale && currentLocaleIso

  if (!hasCurrentLocaleAndIso) {
    return
  }

  meta.push({
    hid: 'i18n-og',
    property: 'og:locale',
    // Replace dash with underscore as defined in spec: language_TERRITORY
    content: hypenToUnderscore(currentLocaleIso)
  })
}

function addAlternateOgLocales(locales: LocaleObject[], currentLocaleIso: string | undefined, meta: MetaAttrs) {
  const localesWithoutCurrent = locales.filter(locale => {
    const localeIso = locale.iso
    return localeIso && localeIso !== currentLocaleIso
  })

  if (localesWithoutCurrent.length) {
    const alternateLocales = localesWithoutCurrent.map(locale => ({
      hid: `i18n-og-alt-${locale.iso}`,
      property: 'og:locale:alternate',
      content: hypenToUnderscore(locale.iso!)
    }))

    meta.push(...alternateLocales)
  }
}

function hypenToUnderscore(str: string) {
  return (str || '').replace(/-/g, '_')
}

function toAbsoluteUrl(urlOrPath: string, baseUrl: string) {
  if (urlOrPath.match(/^https?:\/\//)) {
    return urlOrPath
  }
  return baseUrl + urlOrPath
}
