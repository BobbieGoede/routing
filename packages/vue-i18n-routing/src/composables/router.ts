import { getCurrentInstance, computed, isVue2 } from 'vue-demi'
import VueRouter from 'vue-router3'
import { useRoute as useModernRoute, useRouter as useModernRouter } from 'vue-router'

import type { Route } from 'vue-router3'

const useLegacyRouter = () => {
  const vm = getCurrentInstance()
  if (vm == null) {
    throw new Error(`should be used in setup`)
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (vm as any).$router as VueRouter
}

const useLegacyRoute = () => {
  const vm = getCurrentInstance()
  if (vm == null) {
    throw new Error(`should be used in setup`)
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return computed(() => (vm as any).$route as Route)
}

export function useRoute() {
  return isVue2 ? useLegacyRoute() : /* #__PURE__*/ useModernRoute()
}

export function useRouter() {
  return isVue2 ? useLegacyRouter() : /* #__PURE__*/ useModernRouter()
}
