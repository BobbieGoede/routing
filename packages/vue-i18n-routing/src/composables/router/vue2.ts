import { getCurrentInstance, computed } from 'vue-demi'
import VueRouter from 'vue-router3'

import type { Route } from 'vue-router3'

export const useLegacyRouter = () => {
  const vm = getCurrentInstance()
  if (vm == null) {
    throw new Error(`should be used in setup`)
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (vm as any).$router as VueRouter
}

export const useLegacyRoute = () => {
  const vm = getCurrentInstance()
  if (vm == null) {
    throw new Error(`should be used in setup`)
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return computed(() => (vm as any).$route as Route)
}
