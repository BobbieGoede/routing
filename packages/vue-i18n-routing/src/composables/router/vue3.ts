export { useRoute as useModernRoute, useRouter as useModernRouter } from 'vue-router'
