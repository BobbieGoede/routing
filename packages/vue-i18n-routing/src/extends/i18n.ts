import { isBoolean, isObject } from '@intlify/shared'
import { ref, computed, isVue3, effectScope, isVue2 } from 'vue-demi'
import { resolveBaseUrl, isVueI18n, getComposer } from '../utils'
import { DEFAULT_BASE_URL } from '../constants'
import {
  localePath,
  localeRoute,
  localeLocation,
  switchLocalePath,
  getRouteBaseName,
  resolveRoute,
  localeHead
} from '../compatibles'

import type { App, EffectScope } from 'vue-demi'
import type { I18n, Composer, VueI18n, ExportedGlobalComposer } from '@intlify/vue-i18n-bridge'
import type { I18nRoutingOptions, LocaleObject } from '../types'

type VueI18nExtendOptions = Pick<I18nRoutingOptions, 'baseUrl'> & {
  locales?: string[] | LocaleObject[]
  localeCodes?: string[]
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Vue = any

// eslint-disable-next-line @typescript-eslint/ban-types
function proxyVueInstance(target: Function): Function {
  // `this` is the Vue instance
  return function (this: Vue) {
    return Reflect.apply(
      target,
      {
        getRouteBaseName: this.getRouteBaseName,
        localePath: this.localePath,
        localeRoute: this.localeRoute,
        localeLocation: this.localeLocation,
        resolveRoute: this.resolveRoute,
        switchLocalePath: this.switchLocalePath,
        localeHead: this.localeHead,
        i18n: this.$i18n,
        route: this.$route,
        router: this.$router
      },
      // eslint-disable-next-line prefer-rest-params
      arguments
    )
  }
}

/**
 * An options of Vue I18n Routing Plugin
 */
export interface VueI18nRoutingPluginOptions {
  /**
   * Whether to inject some option APIs style methods into Vue instance
   *
   * @defaultValue `true`
   */
  inject?: boolean
}

export function extendI18n<TI18n extends I18n>(
  i18n: TI18n,
  { locales = [], localeCodes = [], baseUrl = DEFAULT_BASE_URL }: VueI18nExtendOptions = {}
) {
  // TODO: release reactive resources with `effectSceope`

  const orgInstall = i18n.install
  i18n.install = (vue: Vue, ...options: unknown[]) => {
    Reflect.apply(orgInstall, i18n, [vue, ...options])

    const composer = getComposer(i18n)

    // extend global
    extendComposer(composer, { locales, localeCodes, baseUrl })
    if (isVueI18n(i18n.global)) {
      extendVueI18n(i18n.global)
    }

    // extend vue component instance for Vue 3
    const app = vue as App
    // prettier-ignore
    const exported = i18n.mode === 'composition'
      ? isVue3
        ? app.config.globalProperties.$i18n
        : i18n
      : isVue2
        ? i18n
        : null
    if (exported) {
      extendExportedGlobal(exported, composer)
    }

    const pluginOptions = isPluginOptions(options[0]) ? options[0] : { inject: true }
    if (pluginOptions.inject) {
      // extend vue component instance
      vue.mixin({
        methods: {
          resolveRoute: proxyVueInstance(resolveRoute),
          localePath: proxyVueInstance(localePath),
          localeRoute: proxyVueInstance(localeRoute),
          localeLocation: proxyVueInstance(localeLocation),
          switchLocalePath: proxyVueInstance(switchLocalePath),
          getRouteBaseName: proxyVueInstance(getRouteBaseName),
          localeHead: proxyVueInstance(localeHead)
        }
      })
    }
  }
}

function extendComposer(composer: Composer, options: Required<VueI18nExtendOptions>) {
  const { locales, localeCodes, baseUrl } = options

  const _locales = ref<string[] | LocaleObject[]>(locales)
  const _localeCodes = ref<string[]>(localeCodes)

  composer.locales = computed(() => _locales.value)
  composer.localeCodes = computed(() => _localeCodes.value)
  composer.__baseUrl = resolveBaseUrl(baseUrl, {})
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function extendExportedGlobal(exported: any, global: Composer) {
  Object.defineProperty(exported, 'locales', {
    get() {
      return global.locales.value
    }
  })
  Object.defineProperty(exported, 'localeCodes', {
    get() {
      return global.localeCodes.value
    }
  })
  Object.defineProperty(exported, '__baseUrl', {
    get() {
      return global.__baseUrl
    }
  })
}

function extendVueI18n(vueI18n: VueI18n): void {
  const composer = getComposer(vueI18n)
  Object.defineProperty(vueI18n, 'locales', {
    get() {
      return composer.locales.value
    }
  })
  Object.defineProperty(vueI18n, 'localeCodes', {
    get() {
      return composer.localeCodes.value
    }
  })
  Object.defineProperty(vueI18n, '__baseUrl', {
    get() {
      return composer.__baseUrl
    }
  })
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isPluginOptions(options: any): options is VueI18nRoutingPluginOptions {
  return isObject(options) && 'inject' in options && isBoolean(options.inject)
}
