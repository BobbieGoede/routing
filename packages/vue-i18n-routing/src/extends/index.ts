export * from './router'
export type { VueI18nRoutingPluginOptions } from './i18n'
