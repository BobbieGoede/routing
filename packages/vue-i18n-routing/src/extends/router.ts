import VueRouter3 from '@intlify/vue-router-bridge'
import { createRouter as _createRouter } from '@intlify/vue-router-bridge'
import { isString } from '@intlify/shared'
import { isVue2, isVue3 } from 'vue-demi'
import { extendI18n } from './i18n'
import { localizeRoutes } from '../resolve'
import { getLocale, setLocale, getNormalizedLocales } from '../utils'
import {
  DEFAULT_LOCALE,
  DEFAULT_LOCALE_ROUTE_NAME_SUFFIX,
  DEFAULT_ROUTES_NAME_SEPARATOR,
  DEFAULT_STRATEGY,
  DEFAULT_TRAILING_SLASH,
  DEFAULT_DETECTION_DIRECTION,
  DEFAULT_BASE_URL
} from '../constants'

import type {
  Route,
  Router,
  VueRouter,
  RouteLocationNormalizedLoaded,
  RouteLocationNormalized
} from '@intlify/vue-router-bridge'
import type { I18n } from '@intlify/vue-i18n-bridge'
import type { VueI18nRoute, VueI18nRoutingOptions } from '../types'

function getLocalesRegex(localeCodes: string[]) {
  return new RegExp(`^/(${localeCodes.join('|')})(?:/|$)`, 'i')
}

function createLocaleFromRouteGetter(
  localeCodes: string[],
  routesNameSeparator: string,
  defaultLocaleRouteNameSuffix: string
) {
  const localesPattern = `(${localeCodes.join('|')})`
  const defaultSuffixPattern = `(?:${routesNameSeparator}${defaultLocaleRouteNameSuffix})?`
  const regexpName = new RegExp(`${routesNameSeparator}${localesPattern}${defaultSuffixPattern}$`, 'i')
  const regexpPath = getLocalesRegex(localeCodes)

  /**
   * extract locale code from given route:
   * - if route has a name, try to extract locale from it
   * - otherwise, fall back to using the routes'path
   */
  const getLocaleFromRoute = (route: Route | RouteLocationNormalizedLoaded | RouteLocationNormalized): string => {
    // extract from route name
    if (route.name) {
      const name = isString(route.name) ? route.name : route.name.toString()
      const matches = name.match(regexpName)
      if (matches && matches.length > 1) {
        return matches[1]
      }
    } else if (route.path) {
      // Extract from path
      const matches = route.path.match(regexpPath)
      if (matches && matches.length > 1) {
        return matches[1]
      }
    }

    return ''
  }

  return getLocaleFromRoute
}

function asDefaultVueI18nRouterOptions(options: VueI18nRoutingOptions): Required<VueI18nRoutingOptions> {
  options.version = options.version ?? 4
  options.defaultLocale = options.defaultLocale ?? DEFAULT_LOCALE
  options.strategy = options.strategy ?? DEFAULT_STRATEGY
  options.trailingSlash = options.trailingSlash ?? DEFAULT_TRAILING_SLASH
  options.routesNameSeparator = options.routesNameSeparator ?? DEFAULT_ROUTES_NAME_SEPARATOR
  options.defaultLocaleRouteNameSuffix = options.defaultLocaleRouteNameSuffix ?? DEFAULT_LOCALE_ROUTE_NAME_SUFFIX
  options.locales = options.locales ?? []
  options.defaultDirection = options.defaultDirection ?? DEFAULT_DETECTION_DIRECTION
  options.baseUrl = options.baseUrl ?? DEFAULT_BASE_URL
  options.routes = options.routes ?? []
  return options as Required<VueI18nRoutingOptions>
}

/**
 * Create a Vue Router instance
 *
 * @param i18n - A Vue I18n instance, see [Vue I18n API docs](https://vue-i18n.intlify.dev/api/general.html#i18n)
 * @param options - An options, see {@link VueI18nRoutingOptions}
 *
 * @returns A Vue Router instance
 *
 * @remakrs
 * You can create a vue router instance to be used by the Vue app.
 *
 * The routes of the created router instance are handled with i18n routing.
 *
 * At the Vue 2 will return a [Vue Router v3 instance](https://router.vuejs.org/api/#router-construction-options), and at the Vue 3 will return a [Vue Router v4 instance](https://next.router.vuejs.org/api/#createrouter).
 *
 * @public
 */
export function createRouter<Options extends VueI18nRoutingOptions = VueI18nRoutingOptions>(
  i18n: I18n,
  options?: Options
): Options['version'] extends 4 ? Router : VueRouter

export function createRouter(i18n: I18n, options = {} as VueI18nRoutingOptions) {
  const {
    version,
    defaultLocale,
    locales,
    strategy,
    trailingSlash,
    routesNameSeparator,
    defaultLocaleRouteNameSuffix,
    defaultDirection,
    baseUrl,
    routes,
    routeOptionsResolver: optionsResolver
  } = asDefaultVueI18nRouterOptions(options)

  const normalizedLocaleCodes = getNormalizedLocales(locales)
  const localeCodes = normalizedLocaleCodes.map(l => l.code)
  const getLocaleFromRoute = createLocaleFromRouteGetter(localeCodes, routesNameSeparator, defaultLocaleRouteNameSuffix)

  extendI18n(i18n, { locales: normalizedLocaleCodes, baseUrl })

  const localizedRoutes = localizeRoutes(routes as VueI18nRoute[], {
    locales,
    defaultLocale,
    strategy,
    trailingSlash,
    routesNameSeparator,
    defaultLocaleRouteNameSuffix,
    optionsResolver
  })
  options.routes = localizedRoutes as any // eslint-disable-line @typescript-eslint/no-explicit-any

  let router: VueRouter | Router | null = null
  if (isVue3 && version === 4) {
    router = _createRouter(options)
  } else if (isVue2 && version === 3) {
    router = new VueRouter3(options as any) as VueRouter // eslint-disable-line @typescript-eslint/no-explicit-any
  } else {
    // TODO:
    throw new Error('TODO:')
  }

  router.__defaultLocale = defaultLocale
  router.__localeCodes = localeCodes
  router.__strategy = strategy
  router.__trailingSlash = trailingSlash
  router.__routesNameSeparator = routesNameSeparator
  router.__defaultLocaleRouteNameSuffix = defaultLocaleRouteNameSuffix
  router.__defaultDirection = defaultDirection

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const removableGuardListener = router.beforeEach((to, from, next) => {
    // console.log('beforeEach', to, from)
    const currentLocale = getLocale(i18n)
    const finalLocale = getLocaleFromRoute(to) || currentLocale || defaultLocale || ''
    // console.log('currentLocale', currentLocale, 'finalLocale', finalLocale)
    if (currentLocale !== finalLocale) {
      setLocale(i18n, finalLocale)
    }
    next()
  })

  // console.log('create router', router)
  return router
}
