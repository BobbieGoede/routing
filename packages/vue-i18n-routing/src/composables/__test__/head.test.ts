import { createMemoryHistory } from 'vue-router'
import { createI18n } from 'vue-i18n'
import { createRouter } from '../../extends/router'
import { useI18nHead } from '../head'
import { useSetup } from '../../../scripts/vitest'

describe('useI18nHead', () => {
  it('should be worked', async () => {
    const i18n = createI18n({ legacy: false, locale: 'en' })
    const router = createRouter(i18n, {
      version: 4,
      locales: [
        {
          code: 'en',
          iso: 'en-US',
          dir: 'ltr'
        }
      ],
      baseUrl: 'http://localhost:8080',
      routes: [
        { path: '/', name: 'index', component: { template: '<div>index</div>' } },
        { path: '/about', name: 'about', component: { template: '<div>About</div>' } },
        { path: '/:pathMatch(.*)*', name: 'not-found', component: { template: '<div>Not Found</div>' } }
      ],
      history: createMemoryHistory()
    })
    await router.push('/en/about')

    useSetup(() => {
      expect(useI18nHead({ addDirAttribute: true, addSeoAttributes: true })).toMatchSnapshot()
    }, [i18n, router])
  })
})
