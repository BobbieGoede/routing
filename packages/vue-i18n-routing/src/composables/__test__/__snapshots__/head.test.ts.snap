// Vitest Snapshot v1

exports[`useI18nHead > should be worked 1`] = `
{
  "htmlAttrs": {
    "dir": "ltr",
    "lang": "en-US",
  },
  "link": [
    {
      "hid": "i18n-alt-en",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en",
      "rel": "alternate",
    },
    {
      "hid": "i18n-alt-en-US",
      "href": "http://localhost:8080/en/about",
      "hreflang": "en-US",
      "rel": "alternate",
    },
    {
      "hid": "i18n-can",
      "href": "http://localhost:8080/en/about",
      "rel": "canonical",
    },
  ],
  "meta": [
    {
      "content": "en_US",
      "hid": "i18n-og",
      "property": "og:locale",
    },
  ],
}
`;
